File: LeetCode/0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        n_rows, n_cols = len(board), len(board[0])
        traversing = set()

        def dfs(row, col, idx):
            if idx == len(word):
                return True
            if not (0 <= row < n_rows and 0 <= col < n_cols):
                return False
            if (row, col) in traversing:
                return False
            if board[row][col] != word[idx]:
                return False

            traversing.add((row, col))
            for (r, c) in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
              if dfs(row + r, col + c, idx + 1):
                return True
            traversing.remove((row, col))
            return False

        for r in range(n_rows):
          for c in range(n_cols):
            if dfs(r, c, 0):
              return True
        return False

File: LeetCode/0079-word-search/0079-word-search.js
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
const exist = function (board, word) {
    const row = board.length;
    const col = board[0].length;

    const dr = [0, 0, 1, -1];
    const dc = [1, -1, 0, 0];
    const visited = Array.from({ length: row }, () => Array(col).fill(false));
    
    function dfs(r, c, charIdx) {        
        if (charIdx === word.length) return true;
        if (!(0 <= r && r < row && 0 <= c && c < col)) return false;
        if (board[r][c] !== word[charIdx]) return false;
        if (visited[r][c]) return false;

        visited[r][c] = true;
        for (let i = 0; i < 4; i++) {
            const nr = r + dr[i];
            const nc = c + dc[i];
            if (dfs(nr, nc, charIdx + 1)) return true;
        }
        visited[r][c] = false;
        return false;
    }
    

    for (let r = 0; r < row; r++) {
        for (let c = 0; c < col; c++) {
            if (board[r][c] === word[0] && dfs(r, c, 0)) return true;
        }
    }

    return false;
};
